router/proxy/transport.go

```diff
@@ -0,0 +1,201 @@
+package proxy
+
+import (
+	"bufio"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/crypto/nacl/secretbox"
+	"github.com/flynn/flynn/pkg/random"
+)
+
+var (
+	errNoBackends = errors.New("router: no backends available")
+
+	httpTransport = &http.Transport{
+		Dial:                customDial,
+		TLSHandshakeTimeout: 10 * time.Second, // unused, but safer to leave default in place
+	}
+
+	dialer = &net.Dialer{
+		Timeout:   1 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}
+)
+
+// BackendListFunc returns a slice of backend hosts (hostname:port).
+type BackendListFunc func() []string
+
+type transport struct {
+	getBackends BackendListFunc
+
+	stickyCookieKey   [32]byte
+	useStickySessions bool
+}
+
+func (t *transport) getOrderedBackends(stickyBackend string) []string {
+	backends := t.getBackends()
+	shuffle(backends)
+
+	if stickyBackend != "" {
+		swapToFront(backends, stickyBackend)
+	}
+	return backends
+}
+
+func (t *transport) getStickyBackend(req *http.Request) string {
+	if t.useStickySessions {
+		return getStickyCookieBackend(req, t.stickyCookieKey)
+	}
+	return ""
+}
+
+func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend string) {
+	if !t.useStickySessions {
+		return
+	}
+	if backend := res.Request.URL.Host; backend != originalStickyBackend {
+		setStickyCookieBackend(res, backend, t.stickyCookieKey)
+	}
+}
+
+// always sets the response.Request.URL.Host to the last backend that was
+// connected to.
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	stickyBackend := t.getStickyBackend(req)
+	backends := t.getOrderedBackends(stickyBackend)
+	for _, backend := range backends {
+		req.URL.Host = backend
+		res, err := httpTransport.RoundTrip(req)
+		if err == nil {
+			t.setStickyBackend(res, stickyBackend)
+			return res, nil
+		}
+		if _, ok := err.(dialErr); !ok {
+			return nil, err
+		}
+		// retry, maybe log a message about it
+	}
+	return nil, errNoBackends
+}
+
+func (t *transport) Connect(_ net.Addr) (net.Conn, error) {
```

Comment:
Name the method argument `clientAddr` or similar to avoid confusion, I wasn't sure what it was until I looked at usage.